### DISCUSSION TITLE
Is there a way to get routeId in action/loader every request?

### DISCUSSION THREAD

**Discussion by wangyg115** [remix-run/remix#4394] [Q&A]
In my case, i need get matched route id from LoaderFunction/ActionFunction`s args, and i find a way to do this from the docs, but not the idea way.
Use Express Server template create app, in server.js add code like this:
```js
app.all(
  "*",
  process.env.NODE_ENV === "development"
    ? (req, res, next) => {
        purgeRequireCache();

        return createRequestHandler({
          getLoadContext(req, res) {
            //routeid info gos in loader context
            return { routeId: req.query._data || '' };
          },	
          build: require(BUILD_DIR),
          mode: process.env.NODE_ENV,
        })(req, res, next);
      }
    : createRequestHandler({
        getLoadContext(req, res) {
          //routeid info gos in loader context
          return { routeId: req.query._data || '' };
        },	
        build: require(BUILD_DIR),
        mode: process.env.NODE_ENV,
      })
);
```
This work fine at handleDataRequest in remix-server-runtime,  but not work at handleDocumentRequest which has no _data search params. Is there an other way to get route id info i missed up? thanks a lot.

**Comment by sergiodxa** [comment]
> not work at handleDocumentRequest

Because `handleDocumentRequest` trigger many loaders there's no way to have a single route ID.

What are you trying to achieve? If you want to know the route ID if a route in the loader/action the simplest way is hard code it, the route ID is the path of the route file inside the `app` folder without the extension.

So if you have this file `app/routes/__layout/parent/route.name.tsx` the route ID is `routes/__layout/parent/route.name`

**Reply by kiliman** [reply]
Although technically the route_id is the same as the filename, this PR #3970 will let a single route module service multiple route paths. Each path will have a unique id.

I'm not sure if OP wants to know the specific module (route_id) or the one dependent on the path.

```ts
// get routes from `require(BUILD_DIR).routes` in `getLoadContext`
const routeId = getRouteId(context.routes, request)
```

**Reply by sergiodxa** [reply]
The problem is `/articles/123` could use multiple routes, what should be the route ID then? because you could have `routes/__layout/articles.$articleId` and `routes/__layout.tsx` for example.

For SSR there's no way using the URL to know the route ID, you need to know it for the specific route loader which should be unique per route, even with that PR you would still be able to know the route ID of each route because you're defining the custom route IDs, right? So if I say a route file X has the route ID `abc` and `def` I still know it statically and can hard code it.

**Reply by sergiodxa** [reply]
Anyway, my recommendation would be to avoid depending on the route ID, that's an internal Remix thing.

**Reply by wangyg115** [reply]
> Anyway, my recommendation would be to avoid depending on the route ID, that's an internal Remix thing.

In my opinion the route id is not a internal thing already, because `useMatches ` hook exposed route id . we can search for specific data in matching route by route id:
```js
  const matchingRoutes = useMatches();
  const route =  matchingRoutes.find((route) => route.id === id)
  return route?.data;
```

**Reply by wangyg115** [reply]
> > not work at handleDocumentRequest
> 
> Because `handleDocumentRequest` trigger many loaders there's no way to have a single route ID.
> 
> What are you trying to achieve? If you want to know the route ID if a route in the loader/action the simplest way is hard code it, the route ID is the path of the route file inside the `app` folder without the extension.
> 
> So if you have this file `app/routes/__layout/parent/route.name.tsx` the route ID is `routes/__layout/parent/route.name`

Hard code it is not a good way because i want to get routeid from every loader/action function on every route. and i want to know your opinions about the way i figured out, Is it possible to change the code just for one case? thank alot.

**Comment by wangyg115** [comment]
Thanks for your answer, and sorry for didn't make me clear.  I want to make a dispatch function which runs in every Remix action/loader function, and dispatches request to my restful code based on routeId. the dispatch function like this:
```ts
const restDispatchFun = async (arg: DataFunctionArgs) => {
  const routeId = arg.context.routeId as string
  const path = `${routeId.replace('routes','./api')}.js`
  //my  module gos in build/api directory, the directory structure is same as app/routes, 
  // so the dispatch function can dynamic load the module.
  // e.g,  any request to app/routes/index.tsx will be dispatched to api/index.js
  const module = await require(path)
  if( module.default) {
    const fun = model.default as ActionFunction
    return fun(arg)
  }
}
```
I figured out a way to get route id in **every** loader/action function, but it is a hack way for now. Is it passibale to make a PR and merge the code? In my opinion the code is completely harmless.  :)
make change to packages/remix-server-runtime/server.ts 
```ts
async function handleDocumentRequest({
  build,
  loadContext,
  matches,
  request,
  routes,
  serverMode,
}: {
....
   if (matches && isActionRequest(request)) {
    actionMatch = getRequestMatch(url, matches);

    try {
      actionResponse = await callRouteAction({
        // routeId gos here
        loadContext: {routeId: actionMatch.route.id, ...loadContext}, 
        action: actionMatch.route.module.action,
        routeId: actionMatch.route.id,
        params: actionMatch.params,
        request: request,
      });
 ....
   let routeLoaderResults = await Promise.allSettled(
    matchesToLoad.map((match) =>
      match.route.module.loader
        ? callRouteLoader({
            //routeId gos here
            loadContext: {routeId: match.route.id, ...loadContext}, 
            loader: match.route.module.loader,
            routeId: match.route.id,
            params: match.params,
            request: loaderRequest,
          })
        : Promise.resolve(undefined)
    )
  ); 
 ....
} 

```

**Reply by machour** [reply]
@wangyg115 don't hesitate to open a PR, it will get more visibility from maintainers 👍🏼

**Reply by wangyg115** [reply]
@machour  Thanks for your encourage，I‘ll do it.
